### PR TITLE
Sign `task-erred` with `run_id` and reject outdated responses

### LINKED ISSUE #7489: Sign `task-erred` with `run_id` and reject outdated responses
#7463 focused on signing (positive) task results and rejecting outdated data. Following the same logic of only accepting the result current task run, we should also sign `task-erred` with the `run_id` and reject outdated responses and reschedule tasks. cc @fjetter

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #7489

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -2859,7 +2869,9 @@ def _handle_compute_task(self, ev: ComputeTaskEvent) -> RecsInstrs:
                 )
             )
         elif ts.state == "error":
-            instructions.append(TaskErredMsg.from_task(ts, stimulus_id=ev.stimulus_id))
+            instructions.append(
+                TaskErredMsg.from_task(ts, run_id=ev.run_id, stimulus_id=ev.stimulus_id)
+            )
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:2874]: coverage says we never test this use case.
This is not something you did; but would you mind taking the opportunity to add a unit test for it? It's a fairly big blind spot.

File: distributed/tests/test_cancelled_state.py [hunk 2]
```diff
@@ -488,6 +489,9 @@ async def release_all_futures():
         await lock_compute.release()
         await exit_compute.wait()
 
+        while f3.key in b.state.tasks:
+            await asyncio.sleep(0.01)
```
Comment (crusaderky) [inline distributed/tests/test_cancelled_state.py:493]:
```suggestion
        await async_poll_for(lambda: f3.key not in b.state.tasks, timeout=5)
```

**Review by crusaderky — COMMENTED** [review]
Could use an extra test; but on the other hand it could be left to a separate PR

File: distributed/worker_state_machine.py (hunk 1 above)
Comment (fjetter) [reply distributed/worker_state_machine.py:2874]: I'm not even sure if the code here is correct. I would assume that in this case (especially with the task-erred signing) this task should rather be released and recomputed instead of being reported erred.

**Review by fjetter — APPROVED** [review]
Changes LGTM

I agree with @crusaderky that we should write a test for that one case but I suggest to do this in a separate PR. As it is, this PR is improving main and the additional edge case is an increase in scope, particularly since I'm not entirely certain if the current code is even correct.

Considering this is not urgent, I'll leave merging up to @hendrikmakait once he's back

File: distributed/worker_state_machine.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/worker_state_machine.py:2874]: Agreed, I think that in this case as well as the case above for `ts.state == "memory"`, we should probably recompute instead of acting based on the existing result. I will create a follow-up issue.
